plugin/src/main/java/io/jenkins/plugins/remotingkafka/kafkaclient/KafkaProducerClient.java

```diff
@@ -0,0 +1,61 @@
+package io.jenkins.plugins.remotingkafka.kafkaclient;
+
+import org.apache.kafka.clients.producer.KafkaProducer;
+import org.apache.kafka.clients.producer.Producer;
+import org.apache.kafka.clients.producer.ProducerRecord;
+import org.apache.kafka.common.errors.AuthorizationException;
+import org.apache.kafka.common.errors.OutOfOrderSequenceException;
+import org.apache.kafka.common.errors.ProducerFencedException;
+
+import java.util.Properties;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+
+public class KafkaProducerClient {
+    private static final Logger LOGGER = Logger.getLogger(KafkaProducerClient.class.getName());
+    private static KafkaProducerClient instance;
+    private Producer<String, String> producer;
+    private Properties props;
+
+    private KafkaProducerClient() {
+
+    }
+
+    public static KafkaProducerClient getInstance() {
+        if (instance == null) {
+            instance = new KafkaProducerClient();
+        }
+        return instance;
+    }
+
+    public void initProducer(String connectionURL) {
+        if (producer == null) {
+            this.setProps(connectionURL);
+            producer = new KafkaProducer<String, String>(props);
+        }
+    }
+
+    private void setProps(String connectionURL) {
+        props = new Properties();
+        props.put("bootstrap.servers", connectionURL);
+        props.put("transactional.id", "my-transactional-id");
+        props.put("key.serializer", "org.apache.kafka.common.serialization.StringSerializer");
+        props.put("value.serializer", "org.apache.kafka.common.serialization.StringSerializer");
+    }
+
+    public void send(String topic, String key, String value) {
+        producer.initTransactions();
+        try {
+            producer.beginTransaction();
+            producer.send(new ProducerRecord<String, String>(topic, key, value));
+            producer.commitTransaction();
+            LOGGER.info("Sent successfully with topic = " + topic + ", key = " + key + ", value = " + value);
+        } catch (ProducerFencedException | OutOfOrderSequenceException | AuthorizationException e) {
+            LOGGER.log(Level.WARNING, "Producer exception", e);
+            producer.close();
+        } catch (Exception e) {
+            LOGGER.log(Level.WARNING, "Kafka exception", e);
+            producer.abortTransaction();
```

Comment:
will it close the producer?

Reply:
@oleg-nenashev `abortTransaction()` will not close the producer, but `close()` does. From Kafka documentation https://kafka.apache.org/11/javadoc/index.html?org/apache/kafka/clients/producer/KafkaProducer.html, these exceptions cannot be recovered, so the only choice is to close the producer if we get these exceptions.

Reply:
nvm but I don't think I will use transaction for a single message, I will change my code to sending a single message for now.